tracks/bash/exercises/hamming/mentoring.md

```diff
@@ -0,0 +1,127 @@
+### Reasonable Solutions
+
+```bash
+#!/usr/bin/env bash
+main () {
+  (( "$#" == 2 )) || {
+      echo "Usage: hamming.sh <string1> <string2>"
+      exit 1
+  }
+  
+  # Regular vars are easier to read when doing fancy parameter expansion.
+  a=$1 b=$2 
+  
+  (( "${#a}" == "${#b}" )) || {
+      echo "left and right strands must be of equal length"
+      exit 1
+  }
+
+  declare -i count
+  for (( i = 0; i < "${#a}"; i++ )); do
+    [[ "${a:i:1}" == "${b:i:1}" ]] || count+=1
+  done
+
+  printf '%d\n' "$count"
+}
+
+main "$@"
+```
+```bash
+#!/usr/bin/env bash
+main () {
+    if (( $# != 2 )); then 
+        echo "Usage: hamming.sh <string1> <string2>"
+        exit 1
+    fi
+    if (( ${#1} != ${#2} )); then
+        echo "left and right strands must be of equal length"
+        exit 1
+    fi
+    hamming=0
+    for (( i=0; i<${#1}; i++ )); do
+        [[ "${1:$i:1}" != "${2:$i:1}" ]] && (( ++hamming ))
+    done
+    echo "$hamming"
+}
+
+main "$@"
+```
+
+
+### Common Suggestions
+
+* Suggest using a for loop in bash or other [looping constructs](https://www.gnu.org/software/bash/manual/html_node/Looping-Constructs.html#Looping-Constructs) https://www.gnu.org/software/bash/manual/html_node/Looping-Constructs.html#Looping-Constructs
+
+* Suggest using [Parameter Expansion - Substring Extraction](https://www.gnu.org/software/bash/manual/html_node/Shell-Parameter-Expansion.html) to access the index of the string. https://www.gnu.org/software/bash/manual/html_node/Shell-Parameter-Expansion.html
+
+* Suggest creating a counter to gather hammingDistance
+1) let "hammingDistance+=1"
+2) counter=0; (( ++counter ));
+3) counter=0; counter=(( $counter+1 ));
+
+* Suggest Quote the right-hand side of == in [[ ]] to prevent glob matching, the left hand side can be quoted too.
+```
+run bash hamming.sh 'AAA' 'A?A'
+
+✗ expose subtle '[[ $x == $y ]]' bug
```

Comment:
Is the symbol problematic here in a copy and paste situation?

Line 64 is in a "code block" and is suitable for copy and paste to the terminal, but this line is likely not.  The presentation may be confusing as to its intention.

Reply:
```
bash hamming.sh 'AAA' 'A?A'
```

dropped the "run" syntax that is specific to the bats test suite

Reply:
I copied and adapted `bash hamming.sh 'AAA' 'A?A'` from another mentor note

I suppose the correct code snippet would be the following:
x=$1
y=$2
```bash
[[ "${x:index:1}" != "${y:index:1}" ]] 
```

Reply:
added better explination of expose subtle bug test.